### ISSUE TITLE
CustomerData messages do not render

### ISSUE THREAD

**Issue by bassplayer7** [magento/magento2#28692]
### Preconditions (*)

1. Magento 2.3.5-p1

### Steps to reproduce (*)

1. Trigger error message here: `\Magento\CatalogInventory\Model\StockStateProvider::checkQuoteItemQty()` "The fewest you may purchase is %1" by adding something to cart that is less than the minimum product is configured for. For example, a bundled product, with a child product that has "Option Min Qty" of 2 but only entering 1. (Note: it may be necessary to disable Javascript, change the input, and then submit).
2. Watch for the `customer/section/load` response with a `messages` key in the object.

### Expected result (*)

1. A message should show near the header (via `messages.phtml` in the `messages()` area)
2. The `this.messages` should not be so tightly coupled to `customerData` that is clears `this.messages` when it clears `customerData`'s messages (see: `vendor/magento/module-theme/view/frontend/web/js/view/messages.js`)

### Actual result (*)

1. `this.messages` in `vendor/magento/module-theme/view/frontend/web/js/view/messages.js` is cleared with `customerData.set('messages', {});`

More details based on how this runs `vendor/magento/module-theme/view/frontend/web/js/view/messages.js`:
1. `this.messages` default is an empty array – even though it would crash the Knockout if initialized with.
2. `this.initialize()` runs before the Knockout is initialized and `this.messages` is set from `customerData`.
3. If `this.messages().messages` is not empty (and it won't be if there was a message returned), the `customerData` version of the messages is cleared.
5. *Problem:* that `customerData` observable is the same one the messages one references. As a result, clearing it clears the messages.

One possible solution would involve giving messages it's own observable and getting the data from the `customerData` module.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by rahmanQaiser** [comment]
Any solution so far. I'm also getting same error

**Comment by Mkennethsmith** [comment]
I too am finding this error.

**Comment by huynhlehieunam** [comment]
My solution is:
create a new messages.js file in theme to override core messages.js, copy all core code and create some changes in initialize function:
`initialize: function () {
            this._super();

            this.cookieMessages = _.unique($.cookieStorage.get('mage-messages'), 'text');
            var messageClone = customerData.get('messages').extend({
                disposableCustomerData: 'messages'
            });

            this.messages = Object.assign({},messageClone());


            // Force to clean obsolete messages
            if (!_.isEmpty(this.messages.messages)) {
                customerData.set('messages', {});
            }

            $.cookieStorage.set('mage-messages', '');
        },`
also create a new messages.phtml template in your theme with some changes:
`
    <div role="alert" data-bind="foreach: { data: messages.messages, as: 'message' }" class="messages">` 

This solution is not so good. Does anyone have a better solution?

**Comment by mverstegen** [comment]
I'm also facing the same problem, why can post a temporary fix here?

**Comment by loopyNid** [comment]
I'm also facing the same problem, any fix? or temporaryfix?

**Comment by bassplayer7** [comment]
Note for visitors from Google: this is still a bug but it hasn't been fixed in the core.

**Comment by mattijv** [comment]
@engcom-November The easiest way I found to simulate the problem is to modify the `messages.js` file so that you artificially increase it's size. That way you can simulate a case where the component would be downloaded and initialized slowly (in our case the actual cause was mixins targetting the messages component which slowed the initialization enough that the messages would occasionally fail to be displayed).

```diff
--- messages.js	2024-03-20 06:12:14.199277160 +0000
+++ messages.js	2024-03-20 06:14:23.160867401 +0000
@@ -16,6 +16,8 @@
 ], function ($, Component, customerData, _, escaper) {
     'use strict';
 
+    const s = "some very large string, a few MB at least";
+
     return Component.extend({
         defaults: {
             cookieMessages: [],

```

```console
$ ls -lh messages.js 
-rw-r--r-- 1 www-data www-data 3.1M Mar 20 06:16 messages.js

```

Do the test above with the devtools open and caching disabled, so the file needs to be downloaded every time.

I added some logging to the files to highlight the issue. This output happens when the messages.js is unmodified, and messages are displayed as normal.

![Screenshot from 2024-03-20 08-12-50](https://github.com/magento/magento2/assets/1974547/903082a2-f620-45f9-82e9-d1db7cf20da8)

This is the order when I inflated the messages.js size to slow down the loading/initialization of it (and no messages are displayed).

![Screenshot from 2024-03-20 08-11-56](https://github.com/magento/magento2/assets/1974547/88b4c56c-705b-44ae-b287-4be8a6d0c712)

**Comment by engcom-November** [comment]
Hello @mattijv,

Thank you for quick response!

Tried to verify with given modifications, but we are not able to reproduce the issue.
Increased the messages.js file size as can be seen below:
```
-rw-r--r--  1 owner admin   3.5M Mar 21 12:30 messages.js
```
And disabled cache, Still we were able to get the message in the frontend.

Please let us know if we are missing anything.

Thank you

**Comment by engcom-November** [comment]
Hello @mattijv,

As there is no activity on this issue for a long time, we believe the issue has been resolved, hence closing this issue.
Feel free to raise a new issue or reopen this if you need more assistance.

Thank you.

**Comment by mattijv** [comment]
Looks like this has indeed been finally solved (differently) in the 2.4.7 release :partying_face:
